p2panda-rs/src/storage_provider/traits/operation_store.rs

```diff
@@ -49,5 +49,247 @@ pub trait OperationStore<Operation: AsVerifiedOperation> {
+    #[case::delete_operation_many_prev_ops(delete_operation(&random_previous_operations(12)))]
+    #[async_std::test]
+    async fn insert_get_operations(
+        #[case] operation: Operation,
+        #[from(public_key)] author: Author,
+        operation_id: OperationId,
+        document_id: DocumentId,
+        #[from(test_db)]
+        #[future]
+        db: TestStore,
+    ) {
+        let db = db.await;
+        // Construct the storage operation.
+        let operation = VerifiedOperation::new(&author, &operation_id, &operation).unwrap();
+
+        // Insert the doggo operation into the db, returns Ok(true) when succesful.
+        let result = db.store.insert_operation(&operation, &document_id).await;
+        assert!(result.is_ok());
+
+        // Request the previously inserted operation by it's id.
+        let returned_operation = db
+            .store
+            .get_operation_by_id(operation.operation_id())
+            .await
+            .unwrap()
+            .unwrap();
+
+        assert_eq!(returned_operation.public_key(), operation.public_key());
+        assert_eq!(returned_operation.fields(), operation.fields());
+        assert_eq!(returned_operation.operation_id(), operation.operation_id());
+    }
+
+    #[rstest]
+    #[async_std::test]
+    async fn insert_operation_twice(
+        #[from(verified_operation)] verified_operation: VerifiedOperation,
+        document_id: DocumentId,
+        #[from(test_db)]
+        #[future]
+        db: TestStore,
+    ) {
+        let db = db.await;
+
+        assert!(db
+            .store
+            .insert_operation(&verified_operation, &document_id)
+            .await
+            .is_ok());
+
+        assert_eq!(
+            db.store.insert_operation(&verified_operation, &document_id).await.unwrap_err().to_string(),
+            format!("Error occured when inserting an operation with id OperationId(Hash(\"{}\")) into storage", verified_operation.operation_id().as_str())
+        )
+    }
+
+    #[rstest]
+    #[async_std::test]
+    async fn gets_document_by_operation_id(
+        #[from(verified_operation)]
+        #[with(Some(operation_fields(default_fields())), None, None, None, Some(DEFAULT_HASH.parse().unwrap()))]
+        create_operation: VerifiedOperation,
+        #[from(verified_operation)]
+        #[with(Some(operation_fields(default_fields())), Some(DEFAULT_HASH.parse().unwrap()))]
+        update_operation: VerifiedOperation,
+        document_id: DocumentId,
+        #[from(test_db)]
+        #[future]
+        db: TestStore,
+    ) {
+        let db = db.await;
+
+        assert!(db
+            .store
+            .get_document_by_operation_id(create_operation.operation_id())
+            .await
+            .unwrap()
+            .is_none());
+
+        db.store
+            .insert_operation(&create_operation, &document_id)
+            .await
+            .unwrap();
+
+        assert_eq!(
```

Comment:
Wondering why this line is reported as not being hit 🤔

Reply:
Yeh, I don't know either... I've seen this behaviour before too...